lib/adapters/mielecloudservice.js

```diff
@@ -0,0 +1,66 @@
+'use strict';
+
+// Version 1.0.0 of mielecloudservice lib in discovery
+
+const tools = require('../tools.js');
+
+const adapterName = 'mielecloudservice';
+const reIsMieleXGW3000 = /^<\?xml[\s\S]*?<DEVICES>[\s\S]*?\/homebus\/device[\s\S]*?<\/DEVICES>/;
+
+
+function detect(ip, device, options, callback) {
+    // options.log.debug("MieleCloudService => device=" + JSON.stringify(device));
+    // Miele devices may only be detected via ip & mdns - so quick fail if none of these services
+    if ( (device._source !== 'ip') && (device._source !== 'mdns') ) {
+        return callback(null, false, ip);
+    }
+
+    let detected   = false;
+    let deviceName = 'Miele Appliance';
+
+    if (device._source === 'ip') {
+        tools.httpGet('http://' + ip + '/homebus', (err, data) => {
+            if (!err && data && ( reIsMieleXGW3000.test(data) ) ){
```

Comment:
This code is executed async ... so ip detection will never work ...

Reply:
Best isto make the "if detected" code into  a separate function and call from both relevant places ... first wins. so set callback to "null" after called and check efore calling callback if not null

Reply:
The IP-code is copied from Miele-Adapter, as I don't own a XGW3000 gateway.
I'll try to find someone owning one and fix the code - detection of the gateway itself (hopefully via mdns) willl be suffiicent and is maybe already working.
In worst case gateway devices respond to the miele-adapter detection and WIFI-devices respond to MieleCloudService.